core/objects.go

```diff
@@ -254,6 +274,10 @@ type Challenge struct {
 
 	// Used by dns and dvsni challenges
 	Validation *jose.JsonWebSignature `json:"validation,omitempty"`
+
+	// Contains information about URLs used or redirected to and IPs resolved and
+	// used
+	ValidationRecord *ValidationRecord `json:"validationRecord,omitempty"`
```

Comment:
You have way more layers of indirection than you need here.  Just make a single record type (with components like `URL` optional) and make this an array -- it will just have a single entry in the DVSNI case.